File: scripts/build-second-floor.js
#!/usr/bin/env node

import fetch from 'node-fetch';

const LISTENER_URL = 'http://localhost:8765';

async function callCommand(type, params) {
  try {
    const response = await fetch(LISTENER_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ type, params })
    });
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    const result = await response.json();
    if (!result.success) {
      throw new Error(result.error || 'Command failed');
    }
    
    return result;
  } catch (error) {
    console.error(`Error calling ${type}:`, error.message);
    throw error;
  }
}

async function checkFloorAssets() {
  console.log('Checking available floor assets...\n');
  
  const result = await callCommand('asset.list', {
    folder: '/Game/ModularOldTown/Meshes',
    type: 'StaticMesh',
    filter: 'floor'
  });
  
  console.log('Floor assets:');
  result.assets?.forEach(asset => {
    console.log(`  ${asset.name} - ${asset.path}`);
  });
  
  // Also check for 2m tiles or other floor options
  const tileResult = await callCommand('asset.list', {
    folder: '/Game/ModularOldTown/Meshes',
    type: 'StaticMesh',
    filter: '2m'
  });
  
  console.log('\n2m Assets (possible floor tiles):');
  tileResult.assets?.forEach(asset => {
    console.log(`  ${asset.name} - ${asset.path}`);
  });
  
  return { floorAssets: result.assets, tileAssets: tileResult.assets };
}

async function buildSecondFloor() {
  console.log('\n=== Building Second Floor ===\n');
  
  // First check what assets are available
  const { floorAssets, tileAssets } = await checkFloorAssets();
  
  // House dimensions: 10m x 8m, floor at Z=392
  const floorZ = 392;
  const houseX = 10760; // Center X
  const houseY = 690;   // Center Y
  
  // Use 2m tiles in a 5x4 grid
  const tileSize = 200; // 2m = 200 units
  const tilesX = 5;
  const tilesY = 4;
  
  // Find the best floor tile asset
  let floorAssetPath = null;
  
  // Look for a wood floor or generic floor tile
  const possibleFloorAssets = [...(floorAssets || []), ...(tileAssets || [])];
  const floorAsset = possibleFloorAssets.find(asset => 
    asset.name.toLowerCase().includes('floor') || 
    asset.name.toLowerCase().includes('wood') ||
    asset.name.toLowerCase().includes('2m')
  );
  
  if (floorAsset) {
    floorAssetPath = floorAsset.path;
    console.log(`Using floor asset: ${floorAsset.name}`);
  } else {
    console.error('No suitable floor asset found!');
    // Try a fallback - look for any flat surface asset
    const fallbackResult = await callCommand('asset.list', {
      folder: '/Game/ModularOldTown/Meshes',
      type: 'StaticMesh',
      filter: 'wall'
    });
    
    // Use a wall rotated to be horizontal as a floor
    const wallAsset = fallbackResult.assets?.find(asset => 
      asset.name.toLowerCase().includes('2m') ||
      asset.name.toLowerCase().includes('wall')
    );
    
    if (wallAsset) {
      floorAssetPath = wallAsset.path;
      console.log(`Using wall asset as floor (will rotate): ${wallAsset.name}`);
    }
  }
  
  if (!floorAssetPath) {
    console.error('Could not find any suitable asset for floor!');
    return;
  }
  
  console.log('\nPlacing floor tiles...');
  
  // Calculate starting position (bottom-left corner)
  const startX = houseX - (tilesX * tileSize) / 2 + tileSize / 2;
  const startY = houseY - (tilesY * tileSize) / 2 + tileSize / 2;
  
  // Place floor tiles in a grid
  for (let x = 0; x < tilesX; x++) {
    for (let y = 0; y < tilesY; y++) {
      const posX = startX + x * tileSize;
      const posY = startY + y * tileSize;
      
      const actorName = `SecondFloorTile_${x}_${y}`;
      
      // If using a wall as floor, rotate it 90 degrees on pitch
      const rotation = floorAssetPath.includes('wall') ? [0, 90, 0] : [0, 0, 0];
      
      const result = await callCommand('actor.spawn', {
        assetPath: floorAssetPath,
        location: [posX, posY, floorZ],
        rotation: rotation,
        actorName: actorName,
        folderPath: 'Estate/House/SecondFloor/Floor'
      });
      
      if (result.success) {
        console.log(`  Placed ${actorName} at (${posX}, ${posY}, ${floorZ})`);
      }
    }
  }
  
  console.log('\nFloor tiles placed successfully!');
  
  // Now place corner pieces and walls
  await placeSecondFloorWalls(houseX, houseY, floorZ);
  
  // Take screenshots
  await takeScreenshots(houseX, houseY, floorZ);
  
  // Save the level
  console.log('\nSaving level...');
  await callCommand('level.save', {});
  console.log('Level saved!');
}

async function placeSecondFloorWalls(centerX, centerY, floorZ) {
  console.log('\nPlacing second floor walls and corners...');
  
  // Get available wall and corner assets
  const wallResult = await callCommand('asset.list', {
    folder: '/Game/ModularOldTown/Meshes',
    type: 'StaticMesh',
    filter: 'wall'
  });
  
  const cornerResult = await callCommand('asset.list', {
    folder: '/Game/ModularOldTown/Meshes',
    type: 'StaticMesh',
    filter: 'corner'
  });
  
  // Find appropriate assets
  const wallAsset = wallResult.assets?.find(asset => 
    asset.name.toLowerCase().includes('window') &&
    asset.name.toLowerCase().includes('3m')
  );
  
  const cornerAsset = cornerResult.assets?.find(asset => 
    asset.name.toLowerCase().includes('corner') &&
    asset.name.toLowerCase().includes('3m')
  );
  
  if (!wallAsset || !cornerAsset) {
    console.error('Could not find wall or corner assets!');
    return;
  }
  
  console.log(`Using wall: ${wallAsset.name}`);
  console.log(`Using corner: ${cornerAsset.name}`);
  
  // House dimensions
  const width = 1000;  // 10m
  const depth = 800;   // 8m
  const wallHeight = 300; // 3m walls
  
  // Calculate positions
  const halfWidth = width / 2;
  const halfDepth = depth / 2;
  
  // Place corners first
  const corners = [
    { name: 'NE_Corner', x: centerX - halfWidth, y: centerY - halfDepth, yaw: 0 },
    { name: 'SE_Corner', x: centerX + halfWidth, y: centerY - halfDepth, yaw: 90 },
    { name: 'SW_Corner', x: centerX + halfWidth, y: centerY + halfDepth, yaw: 180 },
    { name: 'NW_Corner', x: centerX - halfWidth, y: centerY + halfDepth, yaw: 270 }
  ];
  
  for (const corner of corners) {
    await callCommand('actor.spawn', {
      assetPath: cornerAsset.path,
      location: [corner.x, corner.y, floorZ],
      rotation: [0, 0, corner.yaw],
      actorName: `SecondFloor_${corner.name}`,
      folderPath: 'Estate/House/SecondFloor/Structure'
    });
  }
  
  // Place walls between corners
  // North wall (faces south into building)
  for (let i = 1; i < 3; i++) {
    const x = centerX - halfWidth + 150 + (i - 1) * 300;
    await callCommand('actor.spawn', {
      assetPath: wallAsset.path,
      location: [x, centerY - halfDepth, floorZ],
      rotation: [0, 0, 270], // Faces south
      actorName: `SecondFloor_NorthWall_${i}`,
      folderPath: 'Estate/House/SecondFloor/Structure'
    });
  }
  
  // South wall (faces north into building)
  for (let i = 1; i < 3; i++) {
    const x = centerX - halfWidth + 150 + (i - 1) * 300;
    await callCommand('actor.spawn', {
      assetPath: wallAsset.path,
      location: [x, centerY + halfDepth, floorZ],
      rotation: [0, 0, 90], // Faces north
      actorName: `SecondFloor_SouthWall_${i}`,
      folderPath: 'Estate/House/SecondFloor/Structure'
    });
  }
  
  // East wall (faces west into building)
  for (let i = 1; i < 2; i++) {
    const y = centerY - halfDepth + 150 + (i - 1) * 300;
    await callCommand('actor.spawn', {
      assetPath: wallAsset.path,
      location: [centerX + halfWidth, y, floorZ],
      rotation: [0, 0, 180], // Faces west
      actorName: `SecondFloor_EastWall_${i}`,
      folderPath: 'Estate/House/SecondFloor/Structure'
    });
  }
  
  // West wall (faces east into building)
  for (let i = 1; i < 2; i++) {
    const y = centerY - halfDepth + 150 + (i - 1) * 300;
    await callCommand('actor.spawn', {
      assetPath: wallAsset.path,
      location: [centerX - halfWidth, y, floorZ],
      rotation: [0, 0, 0], // Faces east
      actorName: `SecondFloor_WestWall_${i}`,
      folderPath: 'Estate/House/SecondFloor/Structure'
    });
  }
  
  console.log('Second floor walls and corners placed!');
}

async function takeScreenshots(centerX, centerY, floorZ) {
  console.log('\nTaking screenshots...');
  
  // 1. Perspective view from southwest
  await callCommand('viewport.camera', {
    location: [centerX + 1500, centerY + 1200, floorZ + 800],
    rotation: [0, -30, -135]
  });
  await callCommand('viewport.screenshot', {});
  console.log('  Screenshot 1: Perspective view');
  
  // 2. Top-down view
  await callCommand('viewport.mode', { mode: 'top' });
  await callCommand('viewport.camera', {
    location: [centerX, centerY, floorZ + 2000],
    rotation: [-90, 0, 0]
  });
  await callCommand('viewport.screenshot', {});
  console.log('  Screenshot 2: Top view');
  
  // 3. Wireframe to check alignment
  await callCommand('viewport.render_mode', { mode: 'wireframe' });
  await callCommand('viewport.screenshot', {});
  console.log('  Screenshot 3: Wireframe view');
  
  // Reset to perspective lit mode
  await callCommand('viewport.mode', { mode: 'perspective' });
  await callCommand('viewport.render_mode', { mode: 'lit' });
}

// Run the script
buildSecondFloor().catch(console.error);

File: scripts/build_second_floor.py
#!/usr/bin/env python3
"""Build the second floor of the house in HomeWorld"""

import requests
import json
import time

# Python listener endpoint
BASE_URL = "http://localhost:8765"

def execute_command(command_type, params):
    """Execute a command via the Python listener"""
    payload = {
        "type": command_type,
        "params": params
    }
    
    try:
        response = requests.post(BASE_URL, json=payload, timeout=30)
        response.raise_for_status()
        return response.json()
    except Exception as e:
        print(f"Error executing {command_type}: {e}")
        return None

def main():
    # First, restart the listener to apply rotation fixes
    print("Restarting listener to apply rotation fixes...")
    result = execute_command("python.execute", {
        "code": """
from uemcp_helpers import restart_listener
restart_listener()
"""
    })
    
    # Wait for listener to restart
    time.sleep(3)
    
    # Check available floor assets
    print("\nChecking available floor assets...")
    result = execute_command("asset.list", {
        "path": "/Game/ModularOldTown/Meshes",
        "filter": "SM_Floor",
        "limit": 10
    })
    
    if result and result.get('success'):
        print(f"Found {result.get('totalCount', 0)} floor assets")
        for asset in result.get('assets', []):
            print(f"  - {asset['name']}")
    
    # Get floor asset dimensions
    print("\nGetting floor tile dimensions...")
    floor_2m_info = execute_command("asset.info", {
        "assetPath": "/Game/ModularOldTown/Meshes/SM_Floor_2m"
    })
    
    floor_1m_info = execute_command("asset.info", {
        "assetPath": "/Game/ModularOldTown/Meshes/SM_Floor_1m"
    })
    
    # Calculate floor layout
    # House is 10m x 8m (1000cm x 800cm)
    # Using 2m tiles: 5 tiles x 4 tiles = 20 tiles total
    # Z position for second floor: 392
    
    print("\nPlacing second floor tiles...")
    tile_count = 0
    
    # Place 2m floor tiles in a 5x4 grid
    for row in range(4):  # 4 rows (800cm / 200cm)
        for col in range(5):  # 5 columns (1000cm / 200cm)
            x = 10260 + (col * 200)  # Start at 10260 (west edge)
            y = 290 + (row * 200)    # Start at 290 (north edge)
            z = 392                   # Second floor height
            
            actor_name = f"Floor_F2_{row}_{col}"
            
            result = execute_command("actor.spawn", {
                "assetPath": "/Game/ModularOldTown/Meshes/SM_Floor_2m",
                "location": [x, y, z],
                "rotation": [0, 0, 0],
                "scale": [1, 1, 1],
                "actorName": actor_name,
                "folderPath": "Estate/House/SecondFloor/Floor"
            })
            
            if result and result.get('success'):
                tile_count += 1
                print(f"  Placed {actor_name} at [{x}, {y}, {z}]")
    
    print(f"\nPlaced {tile_count} floor tiles")
    
    # Now place corner pieces for second floor walls
    print("\nPlacing second floor corner pieces...")
    
    corners = [
        {"name": "Corner_F2_NW", "location": [10260, 290, 392], "rotation": [0, 0, 0]},
        {"name": "Corner_F2_NE", "location": [10260, 1090, 392], "rotation": [0, 0, 90]},
        {"name": "Corner_F2_SE", "location": [11260, 1090, 392], "rotation": [0, 0, 180]},
        {"name": "Corner_F2_SW", "location": [11260, 290, 392], "rotation": [0, 0, 270]}
    ]
    
    for corner in corners:
        result = execute_command("actor.spawn", {
            "assetPath": "/Game/ModularOldTown/Meshes/SM_SimpleCorner_A",
            "location": corner["location"],
            "rotation": corner["rotation"],
            "scale": [1, 1, 1],
            "actorName": corner["name"],
            "folderPath": "Estate/House/SecondFloor/Walls"
        })
        
        if result and result.get('success'):
            print(f"  Placed {corner['name']}")
    
    # Place second floor walls with windows
    print("\nPlacing second floor walls...")
    
    # North wall (3 sections with windows)
    for i in range(3):
        x = 10260
        y = 390 + (i * 300)  # Start at 390, space by 300
        
        result = execute_command("actor.spawn", {
            "assetPath": "/Game/ModularOldTown/Meshes/SM_FlatWall_3m_SquareWin",
            "location": [x, y, 392],
            "rotation": [0, 0, 270],  # Face south into building
            "scale": [1, 1, 1],
            "actorName": f"Wall_F2_North_{i+1}",
            "folderPath": "Estate/House/SecondFloor/Walls"
        })
    
    # South wall (2 sections with windows, leaving space for potential balcony)
    for i in range(2):
        x = 11260
        y = 390 + (i * 300)
        
        result = execute_command("actor.spawn", {
            "assetPath": "/Game/ModularOldTown/Meshes/SM_FlatWall_3m_SquareWin",
            "location": [x, y, 392],
            "rotation": [0, 0, 90],  # Face north into building
            "scale": [1, 1, 1],
            "actorName": f"Wall_F2_South_{i+1}",
            "folderPath": "Estate/House/SecondFloor/Walls"
        })
    
    # East wall (2 sections with mix of 2m and 3m)
    result = execute_command("actor.spawn", {
        "assetPath": "/Game/ModularOldTown/Meshes/SM_FlatWall_3m_SquareWin",
        "location": [10560, 1090, 392],
        "rotation": [0, 0, 180],  # Face west into building
        "scale": [1, 1, 1],
        "actorName": "Wall_F2_East_1",
        "folderPath": "Estate/House/SecondFloor/Walls"
    })
    
    result = execute_command("actor.spawn", {
        "assetPath": "/Game/ModularOldTown/Meshes/SM_FlatWall_2m_SquareWin",
        "location": [10860, 1090, 392],
        "rotation": [0, 0, 180],  # Face west into building
        "scale": [1, 1, 1],
        "actorName": "Wall_F2_East_2",
        "folderPath": "Estate/House/SecondFloor/Walls"
    })
    
    # West wall (similar to east)
    result = execute_command("actor.spawn", {
        "assetPath": "/Game/ModularOldTown/Meshes/SM_FlatWall_3m_SquareWin",
        "location": [10560, 290, 392],
        "rotation": [0, 0, 0],  # Face east into building
        "scale": [1, 1, 1],
        "actorName": "Wall_F2_West_1",
        "folderPath": "Estate/House/SecondFloor/Walls"
    })
    
    result = execute_command("actor.spawn", {
        "assetPath": "/Game/ModularOldTown/Meshes/SM_FlatWall_2m_SquareWin",
        "location": [10860, 290, 392],
        "rotation": [0, 0, 0],  # Face east into building
        "scale": [1, 1, 1],
        "actorName": "Wall_F2_West_2",
        "folderPath": "Estate/House/SecondFloor/Walls"
    })
    
    # Check for balcony asset
    print("\nChecking for balcony assets...")
    result = execute_command("asset.list", {
        "path": "/Game/ModularOldTown/Meshes",
        "filter": "Balcony",
        "limit": 5
    })
    
    if result and result.get('totalCount', 0) > 0:
        print("Found balcony assets, placing on south side...")
        # Place balcony on south side
        result = execute_command("actor.spawn", {
            "assetPath": "/Game/ModularOldTown/Meshes/SM_SimpleBalcony_A",
            "location": [11260, 690, 392],  # Center of south wall
            "rotation": [0, 0, 90],  # Face outward (south)
            "scale": [1, 1, 1],
            "actorName": "Balcony_F2_South",
            "folderPath": "Estate/House/SecondFloor/Details"
        })
    
    # Take screenshots to verify
    print("\nTaking verification screenshots...")
    
    # Perspective view
    execute_command("viewport.camera", {
        "location": [9500, 1500, 900],
        "rotation": [-30, -45, 0]
    })
    execute_command("viewport.screenshot", {
        "filename": "house_second_floor_perspective.png"
    })
    
    # Top view
    execute_command("viewport.mode", {"mode": "top"})
    execute_command("viewport.camera", {
        "location": [10760, 690, 1500],
        "rotation": [-90, 0, 0]
    })
    execute_command("viewport.screenshot", {
        "filename": "house_second_floor_top.png"
    })
    
    # Wireframe view
    execute_command("viewport.render_mode", {"mode": "wireframe"})
    execute_command("viewport.screenshot", {
        "filename": "house_second_floor_wireframe.png"
    })
    
    # Save level
    print("\nSaving level...")
    execute_command("level.save", {})
    
    print("\nSecond floor construction complete!")

if __name__ == "__main__":
    main()